cheetah/accelerator/space_charge_kick.py

```diff
@@ -0,0 +1,610 @@
+        )  # Reverse the x and s dimensions
+        green_func_values[
+            :, num_grid_points_x + 1 :, num_grid_points_y + 1 :, num_grid_points_s + 1 :
+        ] = G_values[:, 1:, 1:, 1:].flip(
+            dims=[1, 2, 3]
+        )  # Reverse all dimensions
+
+        return green_func_values
+
+    def _solve_poisson_equation(
+        self, beam: ParticleBeam, cell_size, grid_dimensions
+    ) -> torch.Tensor:  # works only for ParticleBeam at this stage
+        """
+        Solves the Poisson equation for the given charge density, using FFT convolution.
+        """
+        charge_density = self._array_rho(beam, cell_size, grid_dimensions)
+        charge_density_ft = torch.fft.fftn(charge_density, dim=[1, 2, 3])
+        integrated_green_function = self._integrated_green_function(beam, cell_size)
+        integrated_green_function_ft = torch.fft.fftn(
+            integrated_green_function, dim=[1, 2, 3]
+        )
+        potential_ft = charge_density_ft * integrated_green_function_ft
+        potential = (1 / (4 * torch.pi * epsilon_0)) * torch.fft.ifftn(
+            potential_ft, dim=[1, 2, 3]
+        ).real
+
+        # Return the physical potential
+        return potential[
+            :,
+            : charge_density.shape[1] // 2,
+            : charge_density.shape[2] // 2,
+            : charge_density.shape[3] // 2,
+        ]
+
+    def _E_plus_vB_field(
+        self, beam: ParticleBeam, cell_size: torch.Tensor, grid_dimensions: torch.Tensor
+    ) -> torch.Tensor:
+        """
+        Computes the force field from the potential and the particle positions and
+        speeds, as in https://doi.org/10.1063/1.2837054.
+        """
+        inv_cell_size = 1 / cell_size
+        gamma = self._gammaref(beam)
+        igamma2 = 1 / gamma**2 if gamma != 0 else torch.tensor(0.0)
+        potential = self._solve_poisson_equation(beam, cell_size, grid_dimensions)
+
+        grad_x = torch.zeros_like(potential)
+        grad_y = torch.zeros_like(potential)
+        grad_s = torch.zeros_like(potential)
+
+        # Compute the gradients of the potential, using central differences, with 0
+        # boundary conditions.
+        grad_x[:, 1:-1, :, :] = (potential[:, 2:, :, :] - potential[:, :-2, :, :]) * (
+            0.5 * inv_cell_size[:, 0, None, None, None]
+        )
+        grad_y[:, :, 1:-1, :] = (potential[:, :, 2:, :] - potential[:, :, :-2, :]) * (
+            0.5 * inv_cell_size[:, 1, None, None, None]
+        )
+        grad_s[:, :, :, 1:-1] = (potential[:, :, :, 2:] - potential[:, :, :, :-2]) * (
+            0.5 * inv_cell_size[:, 2, None, None, None]
+        )
+
+        # Scale the gradients with lorentz factor
+        grad_x = -igamma2[:, None, None, None] * grad_x
+        grad_y = -igamma2[:, None, None, None] * grad_y
+        grad_s = -igamma2[:, None, None, None] * grad_s
+
+        return grad_x, grad_y, grad_s
+
+    def _cheetah_to_moments(self, beam: ParticleBeam) -> torch.Tensor:
```

Comment:
This function currently doesn't return anything ... it seems to do things inplace 🤔

Reply:
Actually, `moments` could also be a property of `ParticleBeam` I guess?